### ISSUE TITLE
Add a parsed AuxKernel

### ISSUE THREAD

**Issue by dschwen** [idaholab/moose#4807]
We have a few AuxKernels in Marmot that could be trivially replaced by an AuxKernel that takes a function expression.

**Comment by WilkAndy** [comment]
@dschwen , could you teach me why this isn't just FunctionAux ?

a

**Comment by dschwen** [comment]
Because FunxtionAux cannot couple to other aux or non-linear variables.

**Comment by WilkAndy** [comment]
@dschwen , i don't understand your answer.  Do you have a test or some other example?  I think i'm missing something quite important here.  Sorry for being so annoying!

**Comment by dschwen** [comment]
You simply cannot access a variable (non-linear or auxiliary) from a MooseParsedFunction. So if you want a flexible parsed function AuxKernel that makes an AuxVariable a function several non-linear variables, you need the kernel I'm proposing here. My kernel is guaranteed to only evaluate the function at quadrature points. A Moose Function could be evaluated anywhere and can only be a function of x,y,z,y and Postprocessor values.

**Comment by WilkAndy** [comment]
I was praying that was the case, @dschwen .  This is a very valuable addition for me, as i have wanted to do this multiple times, and have hacked something together.  I'm really pleased.

Can a similar technique be used to create a function-of-a-function?  (ie f(g0(x,y,z,t), g1(x,y,z,t),...) )

I really suggest you write a blog about this stuff on mooseframework.org .  i'd have completely missed it had i not been mucking around on github because i was too tired to do any real work.

a

**Comment by dschwen** [comment]
Functions of functions... I hadn't thought about that. Great idea. I'd be trivial to couple in MooseFunction into parsed materials. I might add that in the future. Right now you can declare constants and couple in variables. In the Parsed materials you can couple in material properties as well (and I'm working on automatic differentiation of those). There is no reason why material properties couldn't be coupled into a ParsedAux kernel, other than that I haven't refactored that part of the code. But for consistency this should be made to work in a future PR.

Thinking of Functions, you could use those either through FunctionAux or GenericFunctionMaterials already, by creating an additional auxvar/materialproperty to translate the function.
